src/amo/components/SearchForm/index.js

```diff
@@ -0,0 +1,236 @@
+
+import 'core/css/inc/lib.scss';
+import './styles.scss';
+
+
+export class SearchFormBase extends React.Component {
+  static propTypes = {
+    addonType: PropTypes.string,
+    api: PropTypes.object.isRequired,
+    i18n: PropTypes.object.isRequired,
+    pathname: PropTypes.string.isRequired,
+    query: PropTypes.string.isRequired,
+    dispatch: PropTypes.func.isRequired,
+    errorHandler: PropTypes.object.isRequired,
+    suggestions: PropTypes.arrayOf(PropTypes.shape({
+      name: PropTypes.string.isRequired,
+      url: PropTypes.string.isRequired,
+      iconUrl: PropTypes.string.isRequired,
+    })).isRequired,
+    loadingSuggestions: PropTypes.bool.isRequired,
+    router: PropTypes.object.isRequired,
+    debounce: PropTypes.func.isRequired,
+  }
+
+  static defaultProps = {
+    debounce: defaultDebounce,
+  };
+
+  constructor(props: Object) {
+    super(props);
+
+    this.state = {
+      searchValue: props.query || '',
+    };
+
+    this.handleSuggestionsFetchRequested = props.debounce(
+      this.handleSuggestionsFetchRequested.bind(this),
+      200
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { query } = nextProps;
+
+    if (this.props.query !== query) {
+      this.setState({ searchValue: query || '' });
+    }
+  }
+
+  getSuggestions() {
+    if (this.props.loadingSuggestions) {
+      // 10 is the maximum number of results returned by the API
+      return Array(10).fill({ component: <LoadingText width={60} /> });
+    }
+
+    return this.props.suggestions;
+  }
+
+  goToSearch(query) {
+    const { addonType, api, pathname, router } = this.props;
+    const filters = { query };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    router.push({
+      pathname: `/${api.lang}/${api.clientApp}${pathname}`,
+      query: convertFiltersToQueryParams(filters),
+    });
+  }
+
+  handleSearch = (e) => {
+    e.preventDefault();
+    this.searchInput.blur();
+    this.goToSearch(this.searchInput.value);
+  }
+
+  handleSearchChange = (e) => {
+    this.setState({ searchValue: e.target.value });
+  }
+
+  handleSuggestionsFetchRequested({ value }) {
+    if (!value) {
+      log.debug(`Ignoring suggestions fetch requested because value is not supplied: ${value}`);
+      return;
+    }
+
+    const { addonType, dispatch, errorHandler } = this.props;
+    const filters = { query: value };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    dispatch(autocompleteStart({
+      errorHandlerId: errorHandler.id,
+      filters,
+    }));
+  }
+
+  handleSuggestionsClearRequested = () => {
+    this.props.dispatch(autocompleteCancel());
+  }
+
+  handleSuggestionSelected = (e, { suggestion }) => {
+    e.preventDefault();
+    this.setState({ searchValue: '' }, () => {
+      this.props.router.push(suggestion.url);
+    });
+  }
+
+  renderSuggestion = (suggestion) => {
+    if (this.props.loadingSuggestions) {
```

Comment:
this is a duplication of logic which is unnecessary and also introduces a maintenance burden. I suggest reacting to the data instead, like this:

```js
if (suggestion.component) {
  // The suggestion is already a rendered component.
  return suggestion.component;
}
```